Add docs for get dwarf (#458)
IntelPython/numba-dppy: docs/user_guides/debugging/tools.rst

Additional tools
=================

Consider the following two examples. ``numba_dppy/examples/debug/simple_sum.py``:

.. literalinclude:: ../../../numba_dppy/examples/debug/simple_sum.py
    :lines: 15-
    :linenos:

Example with njit:

.. literalinclude:: ../../../numba_dppy/examples/debug/njit_basic.py
    :lines: 15-
    :linenos:

Getting the DWARF from .elf file
--------------------------------

> If you need to get the DWARF information from a specific kernel, you need to do the following.
> You need to set ``IGC_ShaderDumpEnable``, IGC will write number of dumps into ``/tmp/IntelIGC``.

Reviewer: ```suggestion
If you need to get the DWARF information from a specific kernel, just enable ``IGC_ShaderDumpEnable`` variable. IGC will write number of dumps into ``/tmp/IntelIGC``.
```

.. code-block:: bash

    export IGC_ShaderDumpEnable=1

To read the DWARF of a kernel, we first need a copy of the IGC generated kernel binary.
To do that, run the Python script in a debugger, and set a breakpoint in the kernel:

.. code-block:: bash

    export NUMBA_DPPY_DEBUGINFO=1
    export NUMBA_OPT=1
    gdb-oneapi -q --args python simple_sum.py
    gdb-oneapi -q python 
    (gdb) break simple_sum.py:22     # Assumes the kernel is in file simple_sum.py, at line 22
    (gdb) run

> Once the breakpoint hits, the kernel has been generated and offloaded.

Reviewer: ```suggestion
Once the breakpoint is hit, the kernel has been generated and offloaded.
```

At that point, the IGFX driver (i.e. our debugger driver) has copied the kernel into a file, and saved it at ``/tmp``.

> All files saved at ``/tmp/IntelIGC/pytho_xxx/``

Reviewer: IGFX is a debugger driver?

Reviewer: ```suggestion
At that point, the IGFX driver (i.e. our debugger driver) has copied the kernel into a file, and saved it at ``/tmp``.
All files saved at ``/tmp/IntelIGC/python_xxx/``
```

> Then, to read the DWARF in that kernel binary (elf), use tool llvm-dwarfdump.
> The app outputs the contents of the DWARF.
> So for example, to create a text file DWARF.dump with the contents of the DWARF data:

Reviewer: ```suggestion
Dump generated DWARF from the kernel binary (elf) via llvm-dwarfdump tool:
```

.. code-block:: bash

>     llvm-dwarfdump xxx.elf > DWARF

Reviewer: Putting dump to a file is not connected to the command, it may be confusing
```suggestion
    llvm-dwarfdump xxx.elf
```

Getting the DWARF from Numba assembly (for njit)
--------------------------------------------------

> We need to set Numba environment variable ``NUMBA_DUMP_ASSEMBLY`` that dump the native assembly code of compiled functions.

Reviewer: ```suggestion
Setting Numba environment variable ``NUMBA_DUMP_ASSEMBLY`` dumps the native assembly code of compiled functions.
```

> Then we run script and get info:

Reviewer: It's redundant

.. code-block:: bash

    NUMBA_DUMP_ASSEMBLY=1 python njit-basic.py > njit-basic-asm.txt

Clear file from prints and unrecognized characters (-, =, !, )

> Get compiled from asm:

Reviewer: ```suggestion
Get compiled from assembler:
```
